File: src-tauri/src/commands/characters.rs
use crate::core::models::character::{
    PlayerCharacter, 
    CreateCharacterRequest, 
    UpdateCharacterRequest,
    AddAchievementRequest,
    UpdateRelationshipRequest
};
use crate::services::character_storage::CharacterStorageService;
use crate::utils::error::{DmAssistantError, DmResult, validate_uuid};
use uuid::Uuid;

#[tauri::command]
pub async fn create_character(req: CreateCharacterRequest) -> DmResult<PlayerCharacter> {
    // Validate campaign ID format
    validate_uuid(&req.campaign_id.to_string(), "Campaign")?;
    
    let character = PlayerCharacter::new(req);
    CharacterStorageService::create_character(character)
}

#[tauri::command]
pub async fn get_character(character_id: String) -> DmResult<Option<PlayerCharacter>> {
    let id = validate_uuid(&character_id, "Character")?;
    
    // We need campaign_id to load from file, but for now we'll return an error
    // This will be fixed when we implement proper campaign management
    Err(DmAssistantError::internal("get_character requires campaign context - use get_character_with_campaign instead"))
}

#[tauri::command]
pub async fn get_character_with_campaign(
    campaign_id: String,
    character_id: String
) -> DmResult<Option<PlayerCharacter>> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    
    CharacterStorageService::get_character(&campaign_uuid, &char_uuid)
}

#[tauri::command]
pub async fn get_characters_by_campaign(campaign_id: String) -> DmResult<Vec<PlayerCharacter>> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    CharacterStorageService::get_characters_by_campaign(&campaign_uuid)
}

#[tauri::command]
pub async fn get_active_characters_by_campaign(campaign_id: String) -> DmResult<Vec<PlayerCharacter>> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    CharacterStorageService::get_active_characters_by_campaign(&campaign_uuid)
}

#[tauri::command]
pub async fn update_character(
    character_id: String,
    req: UpdateCharacterRequest,
) -> DmResult<PlayerCharacter> {
    let id = validate_uuid(&character_id, "Character")?;
    
    // Get current character to find campaign_id
    // For now, we'll need campaign_id as parameter
    Err(DmAssistantError::internal("update_character requires campaign context - use update_character_with_campaign instead"))
}

#[tauri::command]
pub async fn update_character_with_campaign(
    campaign_id: String,
    character_id: String,
    req: UpdateCharacterRequest,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    
    CharacterStorageService::modify_character(&campaign_uuid, &char_uuid, |character| {
        character.update(req);
        Ok(())
    })
}

#[tauri::command]
pub async fn delete_character(
    campaign_id: String,
    character_id: String,
) -> DmResult<bool> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    
    CharacterStorageService::delete_character(&campaign_uuid, &char_uuid)
}

#[tauri::command]
pub async fn add_character_achievement(
    campaign_id: String,
    req: AddAchievementRequest,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    
    let character_id = req.character_id; 
    CharacterStorageService::modify_character(&campaign_uuid, &character_id, |character| {
        character.add_achievement(req);
        Ok(())
    })
}

#[tauri::command]
pub async fn remove_character_achievement(
    campaign_id: String,
    character_id: String,
    achievement_id: String,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    let ach_uuid = validate_uuid(&achievement_id, "Achievement")?;
    
    CharacterStorageService::modify_character(&campaign_uuid, &char_uuid, |character| {
        let initial_count = character.achievements.len();
        character.achievements.retain(|a| a.id != ach_uuid);
        
        if character.achievements.len() == initial_count {
            return Err(DmAssistantError::not_found("Achievement", &achievement_id));
        }
        
        Ok(())
    })
}

#[tauri::command]
pub async fn update_character_relationship(
    campaign_id: String,
    req: UpdateRelationshipRequest,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    
    let character_id = req.character_id;
    CharacterStorageService::modify_character(&campaign_uuid, &character_id, |character| {
        character.update_relationship(req);
        Ok(())
    })
}

#[tauri::command]
pub async fn remove_character_relationship(
    campaign_id: String,
    character_id: String,
    npc_id: String,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    let npc_uuid = validate_uuid(&npc_id, "NPC")?;
    
    CharacterStorageService::modify_character(&campaign_uuid, &char_uuid, |character| {
        let initial_count = character.relationships.len();
        character.relationships.retain(|r| r.npc_id != npc_uuid);
        
        if character.relationships.len() == initial_count {
            return Err(DmAssistantError::not_found("Relationship", &format!("Character {} with NPC {}", character_id, npc_id)));
        }
        
        Ok(())
    })
}

#[tauri::command]
pub async fn get_characters_by_npc_relationship(npc_id: String) -> DmResult<Vec<PlayerCharacter>> {
    let npc_uuid = validate_uuid(&npc_id, "NPC")?;
    
    // This requires cross-campaign search, which is not implemented yet
    CharacterStorageService::get_characters_by_npc_relationship(&npc_uuid)
}

#[tauri::command]
pub async fn level_up_character(
    campaign_id: String,
    character_id: String,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    
    CharacterStorageService::modify_character(&campaign_uuid, &char_uuid, |character| {
        if character.level >= 20 {
            return Err(DmAssistantError::character("Character is already at max level (20)"));
        }
        
        character.level += 1;
        Ok(())
    })
}

#[tauri::command]
pub async fn toggle_character_active(
    campaign_id: String,
    character_id: String,
) -> DmResult<PlayerCharacter> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    let char_uuid = validate_uuid(&character_id, "Character")?;
    
    CharacterStorageService::modify_character(&campaign_uuid, &char_uuid, |character| {
        character.is_active = !character.is_active;
        Ok(())
    })
}

// File persistence functions

#[tauri::command]
pub async fn save_characters_to_file(campaign_id: String) -> DmResult<String> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    
    // Create backup before saving
    let backup_path = CharacterStorageService::backup_characters(&campaign_uuid)
        .unwrap_or_else(|_| std::path::PathBuf::from("backup failed"));
    
    Ok(format!("Characters saved successfully. Backup created at: {}", backup_path.display()))
}

#[tauri::command]
pub async fn load_characters_from_file(campaign_id: String) -> DmResult<Vec<PlayerCharacter>> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    CharacterStorageService::get_characters_by_campaign(&campaign_uuid)
}

#[tauri::command]
pub async fn characters_file_exists(campaign_id: String) -> DmResult<bool> {
    let campaign_uuid = validate_uuid(&campaign_id, "Campaign")?;
    Ok(CharacterStorageService::characters_file_exists(&campaign_uuid))
}

File: src/tauri/commands/characters.ts
import { invoke } from "@tauri-apps/api/core";
import {
  PlayerCharacter,
  CreateCharacterRequest,
  UpdateCharacterRequest,
  AddAchievementRequest,
  UpdateRelationshipRequest,
} from "@/core/entities/Character";

/**
 * Typed wrapper for Tauri character commands
 * Provides type safety and error handling for character operations
 */
export class CharacterApi {
  /**
   * Create a new character
   */
  static async create(
    request: CreateCharacterRequest
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>("create_character", {
        req: request,
      });
      return result;
    } catch (error) {
      console.error("Failed to create character:", error);
      throw new Error(`Failed to create character: ${error}`);
    }
  }

  /**
   * Get character by ID (requires campaign context)
   */
  static async getById(
    campaignId: string,
    characterId: string
  ): Promise<PlayerCharacter | null> {
    try {
      const result = await invoke<PlayerCharacter | null>(
        "get_character_with_campaign",
        {
          campaignId,
          characterId,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to get character:", error);
      throw new Error(`Failed to get character: ${error}`);
    }
  }

  /**
   * Get all characters for a campaign
   */
  static async getByCampaign(campaignId: string): Promise<PlayerCharacter[]> {
    try {
      const result = await invoke<PlayerCharacter[]>(
        "get_characters_by_campaign",
        {
          campaignId,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to get characters by campaign:", error);
      throw new Error(`Failed to get characters by campaign: ${error}`);
    }
  }

  /**
   * Get only active characters for a campaign
   */
  static async getActiveByCampaign(
    campaignId: string
  ): Promise<PlayerCharacter[]> {
    try {
      const result = await invoke<PlayerCharacter[]>(
        "get_active_characters_by_campaign",
        {
          campaignId,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to get active characters:", error);
      throw new Error(`Failed to get active characters: ${error}`);
    }
  }

  /**
   * Update character data
   */
  static async update(
    campaignId: string,
    characterId: string,
    request: UpdateCharacterRequest
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>(
        "update_character_with_campaign",
        {
          campaignId,
          characterId,
          req: request,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to update character:", error);
      throw new Error(`Failed to update character: ${error}`);
    }
  }

  /**
   * Delete character
   */
  static async delete(
    campaignId: string,
    characterId: string
  ): Promise<boolean> {
    try {
      const result = await invoke<boolean>("delete_character", {
        campaignId,
        characterId,
      });
      return result;
    } catch (error) {
      console.error("Failed to delete character:", error);
      throw new Error(`Failed to delete character: ${error}`);
    }
  }

  /**
   * Add achievement to character
   */
  static async addAchievement(
    campaignId: string,
    request: AddAchievementRequest
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>(
        "add_character_achievement",
        {
          campaignId,
          req: request,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to add achievement:", error);
      throw new Error(`Failed to add achievement: ${error}`);
    }
  }

  /**
   * Remove achievement from character
   */
  static async removeAchievement(
    campaignId: string,
    characterId: string,
    achievementId: string
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>(
        "remove_character_achievement",
        {
          campaignId,
          characterId,
          achievementId,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to remove achievement:", error);
      throw new Error(`Failed to remove achievement: ${error}`);
    }
  }

  /**
   * Update or create relationship with NPC
   */
  static async updateRelationship(
    campaignId: string,
    request: UpdateRelationshipRequest
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>(
        "update_character_relationship",
        {
          campaignId,
          req: request,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to update relationship:", error);
      throw new Error(`Failed to update relationship: ${error}`);
    }
  }

  /**
   * Remove relationship with NPC
   */
  static async removeRelationship(
    campaignId: string,
    characterId: string,
    npcId: string
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>(
        "remove_character_relationship",
        {
          campaignId,
          characterId,
          npcId,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to remove relationship:", error);
      throw new Error(`Failed to remove relationship: ${error}`);
    }
  }

  /**
   * Get characters that have relationships with specific NPC
   */
  static async getByNpcRelationship(npcId: string): Promise<PlayerCharacter[]> {
    try {
      const result = await invoke<PlayerCharacter[]>(
        "get_characters_by_npc_relationship",
        {
          npcId,
        }
      );
      return result;
    } catch (error) {
      console.error("Failed to get characters by NPC relationship:", error);
      throw new Error(`Failed to get characters by NPC relationship: ${error}`);
    }
  }

  /**
   * Level up character by one level
   */
  static async levelUp(
    campaignId: string,
    characterId: string
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>("level_up_character", {
        campaignId,
        characterId,
      });
      return result;
    } catch (error) {
      console.error("Failed to level up character:", error);
      throw new Error(`Failed to level up character: ${error}`);
    }
  }

  /**
   * Toggle character active status
   */
  static async toggleActive(
    campaignId: string,
    characterId: string
  ): Promise<PlayerCharacter> {
    try {
      const result = await invoke<PlayerCharacter>("toggle_character_active", {
        campaignId,
        characterId,
      });
      return result;
    } catch (error) {
      console.error("Failed to toggle character active status:", error);
      throw new Error(`Failed to toggle character active status: ${error}`);
    }
  }

  /**
   * Save characters to file (placeholder)
   */
  static async saveToFile(campaignId: string): Promise<string> {
    try {
      const result = await invoke<string>("save_characters_to_file", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to save characters to file:", error);
      throw new Error(`Failed to save characters to file: ${error}`);
    }
  }

  /**
   * Load characters from file (placeholder)
   */
  static async loadFromFile(campaignId: string): Promise<PlayerCharacter[]> {
    try {
      const result = await invoke<PlayerCharacter[]>(
        "load_characters_from_file",
        { campaignId }
      );
      return result;
    } catch (error) {
      console.error("Failed to load characters from file:", error);
      throw new Error(`Failed to load characters from file: ${error}`);
    }
  }

  /**
   * Check if characters file exists for campaign
   */
  static async fileExists(campaignId: string): Promise<boolean> {
    try {
      const result = await invoke<boolean>("characters_file_exists", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to check if characters file exists:", error);
      throw new Error(`Failed to check if characters file exists: ${error}`);
    }
  }
}

// Convenience functions for common operations
export const charactersApi = {
  // CRUD operations
  create: CharacterApi.create,
  getById: (campaignId: string, characterId: string) =>
    CharacterApi.getById(campaignId, characterId),
  getByCampaign: CharacterApi.getByCampaign,
  getActive: CharacterApi.getActiveByCampaign,
  update: (
    campaignId: string,
    characterId: string,
    request: UpdateCharacterRequest
  ) => CharacterApi.update(campaignId, characterId, request),
  delete: (campaignId: string, characterId: string) =>
    CharacterApi.delete(campaignId, characterId),

  // Achievement operations
  addAchievement: (campaignId: string, request: AddAchievementRequest) =>
    CharacterApi.addAchievement(campaignId, request),
  removeAchievement: (
    campaignId: string,
    characterId: string,
    achievementId: string
  ) => CharacterApi.removeAchievement(campaignId, characterId, achievementId),

  // Relationship operations
  updateRelationship: (
    campaignId: string,
    request: UpdateRelationshipRequest
  ) => CharacterApi.updateRelationship(campaignId, request),
  removeRelationship: (
    campaignId: string,
    characterId: string,
    npcId: string
  ) => CharacterApi.removeRelationship(campaignId, characterId, npcId),
  getByNpcRelationship: CharacterApi.getByNpcRelationship,

  // Character progression
  levelUp: (campaignId: string, characterId: string) =>
    CharacterApi.levelUp(campaignId, characterId),
  toggleActive: (campaignId: string, characterId: string) =>
    CharacterApi.toggleActive(campaignId, characterId),

  // File operations
  saveToFile: CharacterApi.saveToFile,
  loadFromFile: CharacterApi.loadFromFile,
  fileExists: CharacterApi.fileExists,
};

export default charactersApi;
